### PR TITLE
NOBLE Nonlinear Low-Rank Branches in TransolverBlock FFN (Retry)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

**NOBLE** (arXiv 2603.06492 — \"Nonlinear Low-Rank Branches\") adds a residual branch `σ(x·W_down)·W_up` alongside each existing linear layer in TransolverBlock FFN MLPs, where `W_down` projects to rank r=16 and `W_up` projects back. The activation `σ` is **CosNet**: `cos(ω·x + φ)` with learnable frequency `ω` and phase `φ` per channel. This makes each FFN layer a sum of its original linear map plus a nonlinear low-rank correction — initialized to zero so training starts from baseline-equivalent behavior.

**Why this should improve p_tan:** The tandem pressure distribution has sharp nonlinear dependencies on wake interaction geometry (inter-foil pressure channel, gap/stagger coupling) that a purely linear projection cannot represent efficiently. NOBLE adds dedicated nonlinear capacity to each FFN layer without disrupting the well-conditioned linear path. CosNet's periodic activation is physically motivated: pressure fluctuations in the inter-foil channel are quasi-periodic functions of gap/stagger geometry. Paper reports 1.47x step speedup and 32% fewer steps to convergence at equivalent parameter count.

**Zero-init safety:** W_up is initialized to zeros so the correction branch outputs exactly zero at initialization, training starts from baseline-equivalent behavior.

This idea comes directly from the human research team directives (issue #1926). Note: PR #2204 was assigned previously but the student branch had no commits — this is the true first run.

### Step 1: Add NOBLELinear wrapper class

Add this class near the top of train.py (after imports, before model classes):

```python
class NOBLELinear(nn.Module):
    """Linear layer + low-rank nonlinear branch: y = xW + sigma(xW_down)W_up.
    CosNet activation: sigma(h) = cos(omega*h + phi) with learnable freq/phase.
    Zero-init on W_up ensures correction=0 at step 0 (baseline-equivalent start).
    """
    def __init__(self, in_f, out_f, rank=16, bias=True):
        super().__init__()
        self.linear = nn.Linear(in_f, out_f, bias=bias)
        self.down = nn.Linear(in_f, rank, bias=False)
        self.up = nn.Linear(rank, out_f, bias=False)
        self.omega = nn.Parameter(torch.ones(rank))   # CosNet frequency
        self.phi = nn.Parameter(torch.zeros(rank))    # CosNet phase
        nn.init.zeros_(self.up.weight)                # zero-init: correction starts at 0

    def forward(self, x):
        h = self.down(x)
        h = torch.cos(self.omega * h + self.phi)      # CosNet periodic activation
        return self.linear(x) + self.up(h)
```

### Step 2: Add config flags

Add to the config dataclass:

```python
noble_branches: bool = False   # Enable NOBLE nonlinear low-rank branches in FFN layers
noble_rank: int = 16           # Rank of the low-rank correction branch
```

### Step 3: Replace FFN nn.Linear layers in Physics_Attention_Irregular_Mesh

Find the FFN definition inside Physics_Attention_Irregular_Mesh.__init__ -- it's an nn.Sequential with two nn.Linear layers (the feed-forward after slice attention). When cfg.noble_branches is True, replace each nn.Linear with NOBLELinear:

```python
if cfg.noble_branches:
    self.ffn = nn.Sequential(
        NOBLELinear(n_hidden, n_hidden, rank=cfg.noble_rank),
        nn.GELU(),
        NOBLELinear(n_hidden, n_hidden, rank=cfg.noble_rank),
    )
else:
    # existing FFN definition unchanged
    self.ffn = nn.Sequential(
        nn.Linear(n_hidden, n_hidden),
        nn.GELU(),
        nn.Linear(n_hidden, n_hidden),
    )
```

This applies to ALL 3 TransolverBlocks since each creates its own Physics_Attention_Irregular_Mesh.

### Step 4: LR group for CosNet params (recommended)

The omega and phi parameters benefit from lower LR to prevent divergence. Find where optimizer parameter groups are built and add CosNet params to the attention group (lower LR multiplier):

```python
noble_cosnet_params = [p for n, p in model.named_parameters()
                       if ('omega' in n or 'phi' in n)]
# Add to attention param group (typically 0.1x base LR):
# {'params': noble_cosnet_params, 'lr': cfg.lr * 0.1}
```

If param group code is complex, base LR 2e-4 is also acceptable for CosNet params.

### Step 5: Parse flags in argparse

Add --noble_branches (store_true) and --noble_rank (int, default 16).

### Step 6: Run experiments

**Trial A - rank=16, seed 42:**
```bash
cd cfd_tandemfoil && python train.py --agent nezuko --wandb_name "nezuko/noble-r16-s42" \
  --wandb_group "nezuko/noble-branches-v2" \
  --noble_branches --noble_rank 16 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --seed 42
```

**Trial B - rank=16, seed 73:**
```bash
cd cfd_tandemfoil && python train.py --agent nezuko --wandb_name "nezuko/noble-r16-s73" \
  --wandb_group "nezuko/noble-branches-v2" \
  --noble_branches --noble_rank 16 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --seed 73
```

Report W&B run IDs and final surface MAE for both seeds in a PR comment.
If p_tan < 28.50 for both seeds, also run rank=8 and rank=32 trials to characterize rank sensitivity.

## Baseline

**Current baseline (PR #2184 — DCT Frequency-Weighted Loss, 2-seed avg):**

| Metric | Baseline | Target to beat |
|--------|----------|----------------|
| p_in   | 13.205   | < 13.21        |
| p_oodc | 7.816    | < 7.82         |
| p_tan  | 28.502   | < 28.50        |
| p_re   | 6.453    | < 6.45         |

W&B runs: 6yfv5lio (seed 42, p_tan=28.432), etepxvjc (seed 73, p_tan=28.572)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline-dct-freq" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5
```

### CODE REVIEW

**Comment by tcapelle** [comment]
## Experiment Results — NOBLE Nonlinear Low-Rank Branches

### Runs
| Seed | W&B Run | Epochs | Time |
|------|---------|--------|------|
| 42 | `uxlbo67c` (online) | 132 | 180.9 min |
| 73 | `xphiwlhu` (offline mode — W&B auth expired) | 132 | 180.1 min |

### Surface MAE Results

| Metric | Seed 42 | Seed 73 | **2-seed avg** | Baseline | Δ vs baseline |
|--------|---------|---------|---------------|----------|---------------|
| p_in | 14.59 | 15.00 | **14.80** | 13.21 | +1.59 ↑ worse |
| p_tan | 29.77 | 30.05 | **29.91** | 28.50 | +1.41 ↑ worse |
| p_oodc | 8.93 | 8.55 | **8.74** | 7.82 | +0.92 ↑ worse |
| p_re | 6.92 | 6.87 | **6.90** | 6.45 | +0.45 ↑ worse |

### Result: Did not beat baseline — all metrics regressed.

### Analysis

NOBLE branches (rank=16, CosNet activation) hurt performance across all splits. Hypotheses for why:

1. **Optimization interference**: The `cos(ω·xW_down + φ)` branch introduces trigonometric nonlinearity on top of the GELU FFN. Even with zero-init on `W_up`, the gradient signal through `cos` is oscillatory and may conflict with the well-conditioned updates from the main linear path.

2. **Increased parameter count**: Each TransolverBlock MLP gained ~3 NOBLELinear layers, each adding `2×hidden×rank + 2×rank` = `2×384×16 + 32` ≈ 12K parameters × 3 layers = 36K extra per block. With 3 blocks, that's ~108K extra parameters, increasing capacity without proportional data coverage.

3. **CosNet sensitivity**: Learnable ω initialized to 1.0 may produce high-frequency activations that impede convergence. A better initialization might set ω → 0 (linear regime) and phase φ → π/2 (making cos(0 + π/2) = 0 at init).

4. **Low-rank bottleneck**: rank=16 is 4% of hidden=384. The bottleneck may not capture enough structure to be useful, while still adding optimization noise.

### Key Observations

- The verification pipeline showed NaN surface MAE for seed 73 (offline mode), which is a known issue with the verification code — the training COMPLETE summary printed correct values.
- Both seeds converged normally (no divergence), suggesting the issue is not instability but suboptimal optimization landscape.
- p_in (in-distribution) regressed by ~12%, suggesting the nonlinear branches are harming the core learning task, not just OOD.

### Suggested Follow-ups

1. **Ablate initialization**: Try φ init = π/2 (zeros CosNet output at init) and ω init = 0.1 (slow-varying, near-linear). This might give a smoother learning signal.
2. **Apply to attention only**: Instead of FFN, add NOBLE correction to the Wqkv projection. The Wqkv linear is a natural candidate since attention heads are already low-rank.
3. **Close and move on**: The core idea of adding CosNet nonlinearity to FFN doesn't seem to help this problem — simpler architectures may be preferable.

STUDENT nezuko — NOBLE nonlinear low-rank branches did not beat baseline.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2205 because ADVISOR: Closing — NOBLE nonlinear low-rank branches (CosNet activation, rank=16) is a clear dead end.

**Results vs current baseline (PR #2207, TE Coordinate Frame):**

| Metric | NOBLE 2-seed avg | Baseline | Δ |
|--------|-----------------|----------|---|
| p_in | 14.80 | 12.49 | +18.5% worse |
| p_oodc | 8.74 | 7.62 | +14.7% worse |
| p_tan | 29.91 | 28.52 | +4.9% worse |
| p_re | 6.90 | 6.41 | +7.6% worse |

All four surface MAE metrics regressed significantly. The CosNet periodic activation (`cos(ω·x + φ)`) introduces oscillatory gradients that interfere with the well-conditioned linear FFN path. Even with zero-init on W_up, the trigonometric nonlinearity creates too much optimization noise for the relatively small capacity gain (rank-16 bottleneck = 4% of hidden dim).

Your analysis is solid — the core issue is optimization interference, not capacity. The periodic activation is fundamentally at odds with the smooth gradient landscape the baseline relies on.

Moving on to a fresh direction. Thank you for the thorough analysis and follow-up suggestions..
